README.md
# create_atom <span> | </span> <a href="https://pub.dev/packages/create_atom"><img src="https://img.shields.io/badge/pub.dev-v2.1.0-red?style=flat-square" alt="flutter"/></a> <a href="https://flutter.dev/"><img src="https://img.shields.io/badge/Built with-flutter-blue?style=flat-square" alt="flutter"/></a> <a href="https://dart.dev/"><img src="https://img.shields.io/badge/Language-Dart-purple?style=flat-square" alt="flutter"/></a>

An animated widget that displays atom with a three electrons revolving around a nucleus at its center.

<table>
  <tr>
    <th><img src="https://raw.githubusercontent.com/Abhi011999/create_atom_flutter/master/images/2.0x.gif" alt="2.0x"/></th>
    <th><img src="https://raw.githubusercontent.com/Abhi011999/create_atom_flutter/master/images/1.0x.gif" alt="1.0x"/></th>
    <th><img src="https://raw.githubusercontent.com/Abhi011999/create_atom_flutter/master/images/0.5x.gif" alt="0.5x"/></th>
  </tr>
  <tr>
    <td align="center"><code>size: 200</code></td>
    <td align="center"><code>size: 100</code></td>
    <td align="center"><code>size: 50</code></td>
  </tr>
</table>

## Some Famous logos using Atom

<table>
  <tr>
    <th align="center">Name</th>
    <th align="center">Code</th>
    <th align="center">Logo</th>
  </tr>
  <tr>
    <td align="center"><h2>React Logo</h2></td>
    <td>
      <pre lang="dart">
        Atom(
          size: 200.0,
          orbit1Angle: math.pi / 2,
          orbit2Angle: math.pi / 6,
          orbit3Angle: - math.pi / 6,
          orbitsColor: Color(0xFF00D8FF),
          electronsColor: Color(0xFF00D8FF),
          nucleusColor: Color(0xFF00D8FF),
        ),
        </pre>
    </td>
    <td><img src="https://raw.githubusercontent.com/Abhi011999/create_atom_flutter/master/images/react_logo.gif" alt="react-logo"/></td>
  </tr>
  <tr>
    <td align="center"><h2>Atom Text Editor Logo</h2></td>
    <td>
      <pre lang="dart">
        Container(
          width: 300.0,
          height: 300.0,
          color: Color(0xFF323232),
          child: Atom(
            size: 200.0,
            orbit1Angle: math.pi / 1.9,
            orbit2Angle: math.pi / 5,
            orbit3Angle: -math.pi / 7,
            orbitsColor: Color(0xFFEED9B7),
            electronsColor: Color(0xFFEED9B7),
            nucleusColor: Color(0xFFEED9B7),
          ),
        ),
        </pre>
    </td>
  <td><img src="https://raw.githubusercontent.com/Abhi011999/create_atom_flutter/master/images/atom_editor_logo.gif" alt="atom-editor-logo"/></td>
  </tr>
    <tr>
    <td align="center"><h2>Proton DB Logo</h2></td>
    <td>
      <pre lang="dart">
        Container(
          width: 300.0,
          height: 300.0,
          color: Color(0xFF1A2233),
          child: Atom(
            size: 200.0,
            nucleusRadiusFactor: 2.5,
            orbitsWidthFactor: 6.0,
            orbit1Angle: math.pi / 2,
            orbit2Angle: math.pi / 6,
            orbit3Angle: -math.pi / 6,
            orbitsColor: Color(0xFFF50057),
          ),
        ),
        </pre>
    </td>
  <td><img src="https://raw.githubusercontent.com/Abhi011999/create_atom_flutter/master/images/protonDB_logo.gif" alt="atom-editor-logo"/></td>
  </tr>
</table>

## Features

- Can be scaled upto any size within parent widget boundaries
- Smooth animations with mathematically pre-derived ellipse path equation
- Transparent background for flexible usage with colors
- 3 electron orbits (support for more may come later)
- Orbits angle can be changed
- Orbits color can be changed (support for individual color may come later)
- Orbits width can be changed
- Nucleus size can be changed
- Nucleus color can be changed
- Electrons color can be changed (support for individual color may come later)
- Individual electron's revolution duration can be changed
- Any widget can be added in the center instead of nucleus

## Getting Started

In the `pubspec.yaml` of your flutter project, add the following dependency:

```yaml
dependencies:
  ...
  create_atom: ^2.1.0-nullsafety
```

In your library add the following import:

```dart
import 'package:create_atom/create_atom.dart';
```

After that run `flutter pub get`

For help getting started with Flutter, view the online [documentation](https://flutter.io/).

## Usage

```dart
...
child: Atom(
  size: 100.0,
),
...
```

Yeah, just like that.

- All properties are optional except `size` property.
- You will get a black atom by default.

#### centerWidget property usage -

Assign `centerWidget` with any widget to display that Widget instead of nucleus.

```dart
...
child: Atom(
  size: 100.0,
  centerWidget: Text("At Center"),
),
...
```

**Note:** If both `nucleusColor` and `centerWidget` are set then the `centerWidget` will get preference.

## Properties

#### Definitions -

| Property            | Definition                                 |
| ------------------- | ------------------------------------------ |
| size                | Defines size of the atom's container       |
| nucleusRadiusFactor | Defines radius size factor of nucleus      |
| orbitsWidthFactor   | Defines width factor of the orbits         |
| orbit1angle         | Defines 1st orbit's angle in radians       |
| orbit2angle         | Defines 2nd orbit's angle in radians       |
| orbit3angle         | Defines 3rd orbit's angle in radians       |
| nucleusColor        | Defines Nucleus Color                      |
| orbitsColor         | Defines Orbits Color                       |
| electronsColor      | Defines Electrons Color                    |
| animDuration1       | Defines Animation Duration of 1st electron |
| animDuration2       | Defines Animation Duration of 2nd electron |
| animDuration3       | Defines Animation Duration of 3rd electron |
| centerWidget        | Defines a widget to display at center      |

#### Types and Initial Values -

| Property            | Type       | Initial Value                        |
| ------------------- | ---------- | ------------------------------------ |
| size                | `double`   | `null` (required)                    |
| nucleusRadiusFactor | `double`   | `1.0`                                |
| orbitsWidthFactor   | `double`   | `1.0`                                |
| orbit1angle         | `double`   | `0.0` (Radians), `0.0` (Degrees)     |
| orbit2angle         | `double`   | `pi/3` (Radians), `60.0` (Degrees)   |
| orbit3angle         | `double`   | `-pi/3` (Radians), `-60.0` (Degrees) |
| nucleusColor        | `Color`    | `Colors.black`                       |
| orbitsColor         | `Color`    | `Colors.black`                       |
| electronsColor      | `Color`    | `Colors.black`                       |
| animDuration1       | `Duration` | `Duration(seconds: 1)`               |
| animDuration2       | `Duration` | `Duration(seconds: 2)`               |
| animDuration3       | `Duration` | `Duration(seconds: 3)`               |
| centerWidget        | `Widget`   | `null`                               |

## Example App

### Checkout here - https://abhi011999.github.io/create_atom_flutter/

<img src="https://raw.githubusercontent.com/Abhi011999/create_atom_flutter/master/images/example_random.gif" width="49%" alt="Random Atoms"/> <img src="https://raw.githubusercontent.com/Abhi011999/create_atom_flutter/master/images/example_playground.gif" width="49%" alt="Atoms Playground"/>

**[Code](https://github.com/Abhi011999/create_atom_flutter/blob/master/example/lib/main.dart)**

## Changelog

### Breaking Changes after the nullsafety migration from v2.0.0-nullsafety.0, see changelog for more info -

See [CHANGELOG.md](https://github.com/Abhi011999/create_atom_flutter/blob/master/CHANGELOG.md) for recent changes.

## Contributions

You can be of great help if you want to improve or add anything !

- If you **found a bug**, open an issue.
- If you **have a feature request**, open an issue.
- If you **want to contribute**, submit a pull request.

## License

> Package and all code within licensed under [MIT License](https://github.com/Abhi011999/create_atom_flutter/blob/master/LICENSE).


CHANGELOG.md
## [2.1.0-nullsafety] - 2020-12-20

- #### This version is a stable nullsafety release but dart's nullsafety is still in preview so you have to make changes to your app/package in which you are going to use it.

- #### package -
    - Added two new customization properties - `nucleusRadiusFactor` and `orbitsWidthFactor`
    - One minor nullsafety fix

- #### example app -
    - Updated app with new properties
    - Background color can also be changed now
    - Minor nullsafety fixes

## [2.0.0-nullsafety.0] - 2020-12-16

- #### package -
    - Migrated to NullSafety

- #### example app -
    - Migrated to NullSafety

## [2.0.0] - 2020-06-13

- #### package -
    - Now the Atom's stack uses flutter's painter library implementation to draw the atom and its animations.
    - Painter library uses Skia graphics renderer to render graphical entities, so performance boosted by a factor 10 !
    - Some size and scaling factors slightly reduced.
    - Reduced many widgets in the Atom's build implementation to make it more performant.
    - Heavy code-refactor and reduced LOC.
    - Implemented dual EaseInOut curve in a single curved path using `CatmullRomCurve`.
    - Tried my best to mimic the original EaseInOut curve for both halves of the electron path.
    - You can find more details here - https://api.flutter.dev/flutter/animation/CatmullRomCurve-class.html.

- #### example app -
    - Updated example app with `centerWidget` property and other v2 changes.
    - App demo is now live at - http://abhi011999.github.io/create_atom_flutter/.

## [1.4.0] - 2020-04-25 (BREAKING CHANGE)

- #### package -
    - Replaced `scale` property with `size` property by implementing new logic
      of creating atom stack.
    
        **For users below v1.4.0 -**
        Replace `scale` property with `size` in your corresponding code and set the
        value of `size` in accordance with this example -
        - If suppose `scale` is 1.0 then `size` will be 215.0 * 1.0 = 215
        - If suppose `scale` is 0.3 then `size` will be 215.0 * 0.3 = 64.5
        Because 215.0 was hard-coded value by default.

    - Fixed centerWidget scaling issue, now it scales with the atom
    - Added extensive code documentation
    - Re-factored and re-structured the libraries

- #### example app - 
    - Updated app with the breaking change

## [1.3.0] - 2020-04-21

- #### package -
    - Added a new `centerWidget` property and feature
    - Corrected orbit angles in degrees
    - Fixed orbits scale clipping above 2.0x
    - Electron paths are now dynamically updated when the parent widget updates

- #### example app -
    - Added a new playground mode (web support will be added soon)

## [1.2.0] - 2020-04-10

- Added example app for demonstration
- Updated documentation
- Removed un-used test file (tests can be added in future)

## [1.1.0] - 2020-04-10

- Added documentation and example project.

## [1.0.0] - 2020-04-09

- Initial release, go create some atoms !!

test/createatom_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:create_atom/create_atom.dart';

void main() {
  test('Test Atom', () {
    final atom = Atom();
  });
}

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:atom_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/src/create_atom.dart
// Copyright 2020 Abhishek Dubey. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import 'dart:math' as math;
import 'package:flutter/material.dart';

import 'electrons_anim.dart';

class _NucleusPainter extends CustomPainter {
  _NucleusPainter(this.size, this.color);

  double size;
  Color color;

  @override
  void paint(Canvas c, Size s) {
    Offset nOffset = Offset(s.width, s.height);

    Paint nPaint = Paint()
      ..color = color
      ..style = PaintingStyle.fill
      ..strokeWidth = 1
      ..strokeCap = StrokeCap.round;

    c.drawCircle(nOffset, size, nPaint);
  }

  @override
  bool shouldRepaint(_NucleusPainter old) => false;
}

class _OrbitsPainter extends CustomPainter {
  _OrbitsPainter(this.strokeWidth, this.strokeWidthFactor, this.color);

  double strokeWidth;
  double strokeWidthFactor;
  Color color;

  @override
  void paint(Canvas c, Size s) {
    Offset oOffset = Offset(s.width / 2, s.height / 2);

    Paint oPaint = Paint()
      ..color = color
      ..style = PaintingStyle.stroke
      ..strokeWidth = strokeWidthFactor * (strokeWidth / 3.0);

    c.drawOval(
      Rect.fromCenter(
        center: oOffset,
        height: 2 * oOffset.dx,
        width: oOffset.dy,
      ),
      oPaint,
    );
  }

  @override
  bool shouldRepaint(_OrbitsPainter old) => false;
}

/// This widget creates a basic atom structure.
class Atom extends StatefulWidget {
  /* Dimensions */

  /// Size of atom's box.
  final double size;

  /// Size of atom's nucleus.
  ///
  /// Size factor -> 0.05
  final double nucleusSize;

  /// Radius factor of atom's nucleus.
  ///
  /// Default -> 1.0
  final double nucleusRadiusFactor;

  /// Size of atom's electron.
  ///
  /// Size factor -> 0.035
  final double electronSize;

  /// Width factor for orbits.
  ///
  /// Default -> 1.0
  final double orbitsWidthFactor;

  /* Angles */

  /// First orbit's angle in radians.
  final double orbit1Angle;

  /// Second orbit's angle in radians.
  final double orbit2Angle;

  /// Third orbit's angle in radians.
  final double orbit3Angle;

  /// Color of atom's nucleus.
  final Color nucleusColor;

  /// Color of atom's orbits.
  final Color orbitsColor;

  /// Color of atom electrons.
  final Color electronsColor;

  /* Speed Durations */

  /// First electron's speed duration.
  final Duration animDuration1;

  /// Second electron's speed duration.
  final Duration animDuration2;

  /// Third electron's speed duration.
  final Duration animDuration3;

  /* Widget */

  /// A Widget that get's displayed at center
  /// instead of nucleus according to user's choice.
  ///
  /// Note: If both the [nucleusColor] and [centerWidget]
  /// are set then [centerWidget] gets the preference.
  ///
  /// Scale factor -> 0.005
  final Widget? centerWidget;

  Atom({
    Key? key,
    required this.size,
    this.nucleusRadiusFactor = 1.0,
    this.orbitsWidthFactor = 1.0,
    this.orbit1Angle = 0.0,
    this.orbit2Angle = (1 / 3) * math.pi,
    this.orbit3Angle = (-1 / 3) * math.pi,
    this.nucleusColor = Colors.black,
    this.orbitsColor = Colors.black,
    this.electronsColor = Colors.black,
    this.animDuration1 = const Duration(seconds: 1),
    this.animDuration2 = const Duration(seconds: 2),
    this.animDuration3 = const Duration(seconds: 3),
    this.centerWidget,
  })  : assert(size > 0),
        nucleusSize = 0.05 * nucleusRadiusFactor * size,
        electronSize = 0.035 * size,
        super(key: key);

  @override
  State<StatefulWidget> createState() => _AtomState();
}

/// State creation of atom's widget.
class _AtomState extends State<Atom> {
  Widget _drawOrbit(s, angle) {
    return Transform.rotate(
      angle: angle,
      child: CustomPaint(
        size: s,
        painter: _OrbitsPainter(widget.electronSize, widget.orbitsWidthFactor, widget.orbitsColor),
      ),
    );
  }

  /// Build method implementation of atom's structure.
  ///
  /// Below is the sequence in which the atom's stack
  /// gets created.
  ///
  /// Custom Widget/Nucleus
  ///           🡓
  ///      Orbits Stack
  ///           🡓
  ///      First Orbit
  ///           🡓
  ///      Second Orbit
  ///           🡓
  ///      Third Orbit
  ///           🡓
  ///    Electrons Stack
  ///           🡓
  ///     First Electron
  ///           🡓
  ///     Second Electron
  ///           🡓
  ///     Third Electron
  ///
  /// All stacks are [Center] positioned and [Colors.transparent].
  ///
  Widget build(BuildContext context) {
    Size s = Size(widget.size, widget.size);

    return SizedBox.fromSize(
      size: s,
      child: Stack(
        alignment: Alignment.center,
        children: <Widget>[
          widget.centerWidget == null
              ? CustomPaint(painter: _NucleusPainter(widget.nucleusSize, widget.nucleusColor))
              : Transform.scale(
                  scale: 0.005 * widget.size,
                  child: widget.centerWidget,
                ),
          _drawOrbit(s, widget.orbit1Angle),
          _drawOrbit(s, widget.orbit2Angle),
          _drawOrbit(s, widget.orbit3Angle),
          ElectronsAnim(widget),
        ],
      ),
    );
  }
}


example/lib/main.dart
// Copyright 2020 Abhishek Dubey. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import 'dart:math';
import 'dart:ui';
import 'package:flutter/material.dart';

import 'package:create_atom/create_atom.dart';
import 'package:flutter_colorpicker/flutter_colorpicker.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        brightness: Brightness.dark,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyAtoms(),
    );
  }
}

class MyAtoms extends StatefulWidget {
  @override
  _MyAtomsState createState() => _MyAtomsState();
}

class _MyAtomsState extends State<MyAtoms> {
  static int _currentIndex = 1;

  static final List<Widget> _pages = [
    RandomAtoms(),
    AtomPlayground(),
  ];

  Widget _showText(text) {
    return Text(
      text,
      style: TextStyle(
        fontFamily: "Quicksand",
        fontSize: 100.0,
        color: Colors.white,
      ),
    );
  }

  Widget _bottomNavBar() {
    return BottomNavigationBar(
      currentIndex: _currentIndex,
      backgroundColor: Colors.black,
      unselectedItemColor: Colors.white,
      items: [
        BottomNavigationBarItem(
          icon: Icon(Icons.all_inclusive),
          label: "Random",
        ),
        BottomNavigationBarItem(
          icon: Icon(Icons.play_arrow),
          label: "Playground",
        ),
      ],
      onTap: (int index) => setState(() => _currentIndex = index),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      appBar: PreferredSize(
        preferredSize: Size.fromHeight(150.0),
        child: SizedBox(
          height: 200.0,
          child: Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              _showText("A "),
              _showText("T "),
              Atom(
                size: 75.25,
                orbitsColor: Colors.white,
                nucleusColor: Colors.white,
                electronsColor: Colors.white,
              ),
              _showText(" M"),
            ],
          ),
        ),
      ),
      body: _pages[_currentIndex],
      bottomNavigationBar: _bottomNavBar(),
    );
  }
}

class RandomAtoms extends StatefulWidget {
  @override
  _RandomAtomsState createState() => _RandomAtomsState();
}

class _RandomAtomsState extends State<RandomAtoms> {
  static final _rand = new Random();
  static int? _r, _g, _b;

  int _genRand(int min, int max) => min + _rand.nextInt(max - min);

  _genColor() {
    return Color.fromRGBO(
      _r!,
      _g!,
      _b!,
      1.0,
    );
  }

  _genRandColor() {
    _r = _genRand(0, 255);
    _g = _genRand(0, 255);
    _b = _genRand(0, 255);

    return _genColor();
  }

  Widget _atomGrid() {
    return GridTile(
      child: Atom(
        size: 129,
        orbitsColor: _genRandColor(),
        nucleusColor: _genRandColor(),
        electronsColor: _genRandColor(),
        orbit1Angle: (pi / 180) * _genRand(0, 361),
        orbit2Angle: (pi / 180) * _genRand(0, 361),
        orbit3Angle: (pi / 180) * _genRand(0, 361),
        animDuration1: Duration(milliseconds: _genRand(250, 5000)),
        animDuration2: Duration(milliseconds: _genRand(250, 5000)),
        animDuration3: Duration(milliseconds: _genRand(250, 5000)),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      body: GridView.count(
        crossAxisCount: 3,
        padding: EdgeInsets.all(4.0),
        children: <Widget>[
          _atomGrid(),
          _atomGrid(),
          _atomGrid(),
          _atomGrid(),
          _atomGrid(),
          _atomGrid(),
          _atomGrid(),
          _atomGrid(),
          _atomGrid(),
        ],
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () => setState(() {}),
        child: Icon(Icons.refresh),
        tooltip: "Randomize",
      ),
    );
  }
}

class AtomPlayground extends StatefulWidget {
  @override
  _AtomPlaygroundState createState() => _AtomPlaygroundState();
}

class _AtomPlaygroundState extends State<AtomPlayground> {
  double _scale = 200;
  double _nRadiusFactor = 1.0;
  double _oWidthFactor = 1.0;
  double _oAngle1 = 0.0;
  double _oAngle2 = 1.047198;
  double _oAngle3 = 5.235988;
  Color _nColor = Color(0xffffffff);
  Color _oColor = Color(0xffffffff);
  Color _eColor = Color(0xffffffff);
  Color _bgColor = Color(0xff1a1a1a);
  double _eSpeed1 = 1000.0;
  double _eSpeed2 = 2000.0;
  double _eSpeed3 = 3000.0;
  Duration _eDur1 = Duration(milliseconds: 1000);
  Duration _eDur2 = Duration(milliseconds: 2000);
  Duration _eDur3 = Duration(milliseconds: 3000);
  Widget? _centerW;

  int? _radioButton;

  void _changeNColor(Color colorNext) {
    setState(() => _nColor = colorNext);
  }

  void _changeOColor(Color colorNext) {
    setState(() => _oColor = colorNext);
  }

  void _changeEColor(Color colorNext) {
    setState(() => _eColor = colorNext);
  }

  void _changeBGColor(Color colorNext) {
    setState(() => _bgColor = colorNext);
  }

  _colorPicker(colorValue, colorFunction) {
    return showDialog(
        context: context,
        builder: (BuildContext context) {
          return AlertDialog(
            content: SingleChildScrollView(
              child: ColorPicker(
                pickerColor: colorValue,
                onColorChanged: colorFunction,
              ),
            ),
          );
        });
  }

  Widget _scaleSlider() {
    return Wrap(
      alignment: WrapAlignment.center,
      crossAxisAlignment: WrapCrossAlignment.center,
      runSpacing: 10.0,
      children: [
        Text(
          "Scale :",
          style: TextStyle(color: Colors.white70),
        ),
        Container(
          width: 350.0,
          child: Slider(
            min: 21.5,
            max: 430,
            value: _scale,
            onChanged: (val) => setState(() => _scale = val),
          ),
        ),
      ],
    );
  }

  Widget _nRadiusSlider() {
    return Wrap(
      alignment: WrapAlignment.center,
      crossAxisAlignment: WrapCrossAlignment.center,
      runSpacing: 10.0,
      children: [
        Text(
          "Nucleus Radius Factor :",
          style: TextStyle(color: Colors.white70),
        ),
        Container(
          width: 350.0,
          child: Slider(
            min: 0.1,
            max: 10.0,
            value: _nRadiusFactor,
            onChanged: (val) => setState(() => _nRadiusFactor = val),
          ),
        ),
      ],
    );
  }

  Widget _oWidthSlider() {
    return Wrap(
      alignment: WrapAlignment.center,
      crossAxisAlignment: WrapCrossAlignment.center,
      runSpacing: 10.0,
      children: [
        Text(
          "Orbits Width Factor :",
          style: TextStyle(color: Colors.white70),
        ),
        Container(
          width: 350.0,
          child: Slider(
            min: 0.1,
            max: 10.0,
            value: _oWidthFactor,
            onChanged: (val) => setState(() => _oWidthFactor = val),
          ),
        ),
      ],
    );
  }

  Widget _oAngleSlider1() {
    return Wrap(
      alignment: WrapAlignment.center,
      crossAxisAlignment: WrapCrossAlignment.center,
      runSpacing: 10.0,
      children: [
        Text(
          "Orbit 1 :",
          style: TextStyle(color: Colors.white70),
        ),
        Container(
          width: 350.0,
          child: Slider(
            min: 0.0,
            max: 2 * pi,
            divisions: 360,
            value: _oAngle1,
            onChanged: (val) => setState(() => _oAngle1 = val),
          ),
        ),
      ],
    );
  }

  Widget _oAngleSlider2() {
    return Wrap(
      alignment: WrapAlignment.center,
      crossAxisAlignment: WrapCrossAlignment.center,
      runSpacing: 10.0,
      children: [
        Text(
          "Orbit 2 :",
          style: TextStyle(color: Colors.white70),
        ),
        Container(
          width: 350.0,
          child: Slider(
            min: 0.0,
            max: 2 * pi,
            divisions: 360,
            value: _oAngle2,
            onChanged: (val) => setState(() => _oAngle2 = val),
          ),
        ),
      ],
    );
  }

  Widget _oAngleSlider3() {
    return Wrap(
      alignment: WrapAlignment.center,
      crossAxisAlignment: WrapCrossAlignment.center,
      runSpacing: 10.0,
      children: [
        Text(
          "Orbit 3 :",
          style: TextStyle(color: Colors.white70),
        ),
        Container(
          width: 350.0,
          child: Slider(
            min: 0.0,
            max: 2 * pi,
            divisions: 360,
            value: _oAngle3,
            onChanged: (val) => setState(() => _oAngle3 = val),
          ),
        ),
      ],
    );
  }

  Widget _colorSelectors() {
    return Padding(
      padding: const EdgeInsets.fromLTRB(0.0, 10.0, 0.0, 10.0),
      child: Wrap(
        alignment: WrapAlignment.center,
        crossAxisAlignment: WrapCrossAlignment.center,
        runSpacing: 10.0,
        children: [
          Text(
            "Nucleus Color :",
            style: TextStyle(color: Colors.white70),
          ),
          VerticalDivider(),
          GestureDetector(
            onTap: () {
              _colorPicker(_nColor, _changeNColor);
            },
            child: Container(
              width: 30.0,
              height: 30.0,
              decoration: BoxDecoration(color: _nColor, shape: BoxShape.circle, border: Border.all(color: Colors.white)),
            ),
          ),
          VerticalDivider(),
          Text(
            "Orbits Color :",
            style: TextStyle(color: Colors.white70),
          ),
          VerticalDivider(),
          GestureDetector(
            onTap: () {
              _colorPicker(_oColor, _changeOColor);
            },
            child: Container(
              width: 30.0,
              height: 30.0,
              decoration: BoxDecoration(color: _oColor, shape: BoxShape.circle, border: Border.all(color: Colors.white)),
            ),
          ),
          VerticalDivider(),
          Text(
            "Electrons Color :",
            style: TextStyle(color: Colors.white70),
          ),
          VerticalDivider(),
          GestureDetector(
            onTap: () {
              _colorPicker(_eColor, _changeEColor);
            },
            child: Container(
              width: 30.0,
              height: 30.0,
              decoration: BoxDecoration(color: _eColor, shape: BoxShape.circle, border: Border.all(color: Colors.white)),
            ),
          ),
          VerticalDivider(),
          Text(
            "Backgroud Color :",
            style: TextStyle(color: Colors.white70),
          ),
          VerticalDivider(),
          GestureDetector(
            onTap: () {
              _colorPicker(_bgColor, _changeBGColor);
            },
            child: Container(
              width: 30.0,
              height: 30.0,
              decoration: BoxDecoration(color: _bgColor, shape: BoxShape.circle, border: Border.all(color: Colors.white)),
            ),
          ),
        ],
      ),
    );
  }

  Widget _eSpeedSlider1() {
    return Wrap(
      alignment: WrapAlignment.center,
      crossAxisAlignment: WrapCrossAlignment.center,
      runSpacing: 10.0,
      children: [
        Text(
          "ESpeed 1 :",
          style: TextStyle(color: Colors.white70),
        ),
        Container(
          width: 350.0,
          child: Slider.adaptive(
            min: 50,
            max: 6000,
            value: _eSpeed1,
            onChanged: (val) => setState(() {
              _eDur1 = Duration(milliseconds: val.toInt());
              _eSpeed1 = val;
            }),
          ),
        ),
      ],
    );
  }

  Widget _eSpeedSlider2() {
    return Wrap(
      alignment: WrapAlignment.center,
      crossAxisAlignment: WrapCrossAlignment.center,
      runSpacing: 10.0,
      children: [
        Text(
          "ESpeed 2 :",
          style: TextStyle(color: Colors.white70),
        ),
        Container(
          width: 350.0,
          child: Slider(
            min: 50,
            max: 6000,
            value: _eSpeed2,
            onChanged: (val) => setState(() {
              _eDur2 = Duration(milliseconds: val.toInt());
              _eSpeed2 = val;
            }),
          ),
        ),
      ],
    );
  }

  Widget _eSpeedSlider3() {
    return Wrap(
      alignment: WrapAlignment.center,
      crossAxisAlignment: WrapCrossAlignment.center,
      runSpacing: 10.0,
      children: [
        Text(
          "ESpeed 3 :",
          style: TextStyle(color: Colors.white70),
        ),
        Container(
          width: 350.0,
          child: Slider(
            min: 50,
            max: 6000,
            value: _eSpeed3,
            onChanged: (val) => setState(() {
              _eDur3 = Duration(milliseconds: val.toInt());
              _eSpeed3 = val;
            }),
          ),
        ),
      ],
    );
  }

  Widget _centerWidgetBar() {
    return Wrap(
      alignment: WrapAlignment.center,
      crossAxisAlignment: WrapCrossAlignment.center,
      children: <Widget>[
        Text(
          "Center Widget :",
          style: TextStyle(color: Colors.white70),
          textAlign: TextAlign.justify,
          overflow: TextOverflow.ellipsis,
        ),
        VerticalDivider(),
        Radio(
          value: 1,
          groupValue: _radioButton,
          onChanged: (dynamic val) {
            setState(() {
              _centerW = null;
              _radioButton = val;
            });
          },
        ),
        Text(
          "No",
          style: TextStyle(color: Colors.white70),
        ),
        VerticalDivider(width: 5.0),
        Radio(
          value: 2,
          groupValue: _radioButton,
          onChanged: (dynamic val) {
            setState(() {
              _centerW = Atom(
                size: 50,
                nucleusColor: _nColor,
                orbitsColor: _oColor,
                electronsColor: _eColor,
              );
              _radioButton = val;
            });
          },
        ),
        Text(
          "Atom",
          style: TextStyle(color: Colors.white70),
        ),
        VerticalDivider(width: 5.0),
        Radio(
          value: 3,
          groupValue: _radioButton,
          onChanged: (dynamic val) {
            setState(() {
              _centerW = Text(
                "Atom",
                style: TextStyle(color: _nColor),
              );
              _radioButton = val;
            });
          },
        ),
        Text(
          "Text",
          style: TextStyle(color: Colors.white70),
        ),
      ],
    );
  }

  void initState() {
    super.initState();
    _radioButton = 1;
  }

  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: _bgColor,
      body: Center(
        child: Padding(
          padding: EdgeInsets.fromLTRB(10.0, 10.0, 10.0, 10.0),
          child: Stack(
            children: [
              Align(
                alignment: Alignment.topCenter,
                child: Atom(
                  size: _scale,
                  nucleusRadiusFactor: _nRadiusFactor,
                  orbitsWidthFactor: _oWidthFactor,
                  orbit1Angle: _oAngle1,
                  orbit2Angle: _oAngle2,
                  orbit3Angle: _oAngle3,
                  nucleusColor: _nColor,
                  orbitsColor: _oColor,
                  electronsColor: _eColor,
                  animDuration1: _eDur1,
                  animDuration2: _eDur2,
                  animDuration3: _eDur3,
                  centerWidget: _centerW,
                ),
              ),
              Align(
                alignment: Alignment.bottomCenter,
                child: ClipRect(
                  child: BackdropFilter(
                    filter: ImageFilter.blur(sigmaX: 10.0, sigmaY: 10.0),
                    child: Container(
                      height: 200,
                      decoration: BoxDecoration(
                        color: Colors.black.withOpacity(0.5),
                        borderRadius: BorderRadius.circular(10.0),
                      ),
                      child: Scrollbar(
                        child: SingleChildScrollView(
                          padding: EdgeInsets.fromLTRB(10.0, 20.0, 10.0, 10.0),
                          physics: BouncingScrollPhysics(),
                          child: Column(
                            children: [
                              _scaleSlider(),
                              Divider(thickness: 2.0),
                              _nRadiusSlider(),
                              _oWidthSlider(),
                              Divider(thickness: 2.0),
                              _oAngleSlider1(),
                              _oAngleSlider2(),
                              _oAngleSlider3(),
                              Divider(thickness: 2.0),
                              _colorSelectors(),
                              Divider(thickness: 2.0),
                              _eSpeedSlider1(),
                              _eSpeedSlider2(),
                              _eSpeedSlider3(),
                              Divider(thickness: 2.0),
                              _centerWidgetBar(),
                            ],
                          ),
                        ),
                      ),
                    ),
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
